### PR TITLE
Push Notifications API

### PR DESCRIPTION
**Pull request by gabrielsaruhashi**
TO-DO
- Call Reminder Notif (cron job) 
- Facility bulk cancel notif (e.g lockdown)
- Create endpoints to update notifications to 'read'

### CODE REVIEW

File: src/notifications/entities/Device.entity.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+// Note: currently only outside users will have devices registered on the DB
+// However, in the future, we hope to give inc people individual devices to use Ameelio
+@Entity({ name: 'Device' })
+class Device extends BaseEntity {
+  @Column({ unique: true })
+  malanUserId: string;
+
+  @Column()
+  fcmToken: string;
```
Comment (mpekala23) [inline src/notifications/entities/Device.entity.ts:13]: i would call this device token, `fcmToken` is android specific language, on ios it's `apnToken`

File: src/notifications/processors/notifications.processor.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    await this.notificationsService.create(job.data.notification);
+
+    await Promise.allSettled([
+      this.dispatchToIOSDevices(notification, devices),
+      this.dispatchToAndroidDevices(notification, devices),
+    ]);
+
+    await this.notificationsService.update(notification.id, {
+      sentAt: new Date(),
+    });
```
Comment (mpekala23) [inline src/notifications/processors/notifications.processor.ts:45]: idk how accurate we care about this info being, but the `dispatchTo` calls could actually take a while to resolve and be fairly different for different users depending on the order (especially with a lot of users). definitely doesn't matter now, probably won't matter later but if we want to be able to get more granular info on response times from notif sent time might be worth retooling this

File: src/notifications/processors/notifications.processor.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    const cleanAndroidTokens = devices
+      .filter((device) => device.type === DeviceType.ANDROID)
+      .map((device) => device.fcmToken);
+
+    return await Promise.all(
+      cleanAndroidTokens.map((token) => {
+        const body = { ...blankBody, to: token };
+
+        return fetch('https://fcm.googleapis.com/fcm/send', {
```
Comment (mpekala23) [inline src/notifications/processors/notifications.processor.ts:106]: idk if this works for you (not sure if you've tested), but this gave me errors until i started using axios. Not entirely sure why, i might've been missing something obvious but just a heads up.

File: src/users/users.controller.ts [hunk 4]
```diff
@@ -216,4 +217,34 @@ export default class UsersController {
...
+    description: 'Should be used when user logs out of their account.',
+  })
+  async deactivateUserDevice(
+    @Body() patchDeviceDto: PatchDeviceDto,
+    @Req() { user }: RequestWithUser,
+  ): Promise<Device> {
+    return await this.usersService.genDeactivateUserDevice({
+      ...patchDeviceDto,
+      malanUserId: user.malanId,
+    });
```
Comment (mpekala23) [inline src/users/users.controller.ts:248]: 💯💯💯💯💯

File: src/users/users.service.ts [hunk 5]
```diff
@@ -82,14 +82,32 @@ export default class UsersService extends CrudService<User> {
...
+    const inmates = call.inmateParticipants;
+
+    inmates.forEach((inmate) => {
+      this.notificationsService.createNotification({
+        title: `${inmate.firstName} scheduled a call with you U+1F4C5`,
+        body: `${format(
+          call.scheduledStart,
+          'E..EEEMMM d, yyyy @ hh:mm -',
+        )} ${format(call.scheduledEnd, 'hh:mm O...OOO')}`,
+        recipientMalanId: inmate.malanId, // change this
```
Comment (mpekala23) [inline src/users/users.service.ts:106]: do we need to change? idk

File: src/users/users.service.ts [hunk 6]
```diff
@@ -117,7 +135,18 @@ export default class UsersService extends CrudService<User> {
...
+    const inmates = call.inmateParticipants;
+
+    inmates.forEach((inmate) => {
+      this.notificationsService.createNotification({
+        title: `${inmate.firstName} cancelled a call with you U+1F4C5`,
+        body: `${format(
+          call.scheduledStart,
+          'E..EEEMMM d, yyyy @ hh:mm -',
+        )} ${format(call.scheduledEnd, 'hh:mm O...OOO')}`,
+        recipientMalanId: inmate.malanId, // change this
```
Comment (mpekala23) [inline src/users/users.service.ts:147]: same

**Review by mpekala23 — APPROVED** [review]
WOWOWOW 🎉🎉🎉🎉🎉 can't believe how quickly you did this sheesh, looks great! a couple nits / questions that I think are worth addressing, but marked as approved in case you want to get moving

File: src/facilities/facilities.service.ts [hunk 7]
```diff
@@ -118,6 +121,49 @@ export default class FacilitiesService extends CrudService<Facility> {
...
+        : 'U+2639';
+    const notifStatus =
+      updatedConnection.status === ConnectionStatus.ACTIVE
+        ? 'approved'
+        : 'rejected';
+
+    const notifBody = ConnectionStatus.ACTIVE
+      ? 'Tap to view their profile'
+      : updatedConnection.statusDetails ||
+        'Contact the facility for more details';
```
[the remark this answers is not in the document]
Comment (gabrielsaruhashi) [reply src/facilities/facilities.service.ts:137]: oh that's smart --- i'll open a task for this

File: src/inmates/inmates.service.ts [hunk 8]
```diff
@@ -124,6 +132,17 @@ export default class InmatesService extends CrudService<Inmate> {
...
       ...dto,
     });
+
+    call.userParticipants.forEach((user) => {
+      this.notificationsService.createNotification({
+        title: `${user.firstName} cancelled a call with you U+1F4C5`,
+        body: `${format(
+          call.scheduledStart,
+          'E..EEEMMM d, yyyy @ hh:mm -',
+        )} ${format(call.scheduledEnd, 'hh:mm O...OOO')}`,
```
[the remark this answers is not in the document]
Comment (gabrielsaruhashi) [reply src/inmates/inmates.service.ts:142]: oh that's cool - can u point me to the implementation?

File: src/notifications/processors/notifications.processor.ts (hunk 3 above)
Comment (gabrielsaruhashi) [reply src/notifications/processors/notifications.processor.ts:117]: gotta test on android hmm will keep it for now

File: src/notifications/processors/notifications.processor.ts (hunk 2 above)
Comment (gabrielsaruhashi) [reply src/notifications/processors/notifications.processor.ts:39]: oh where do u think we should place this call?

**Review by lspdrz — APPROVED** [review]

**Pull request merged**